### ISSUE TITLE
Play Separated Video & Audio Files

### ISSUE THREAD

**Issue by namdevel** [shiyiya/oplayer#78]
hello, how to play separated video and audio files

example i have 2 file
1. video without audio sound
2. audio sound of video

**Comment by shiyiya** [comment]
@namdevel 

You can try this, but why not use ffmpeg to merge them?

```ts
class SeparatedAudio implements PlayerPlugin {
  key = 'hello'
  name = 'oplayer-plugin-hello'
  version = __VERSION__

  private $audio: HTMLAudioElement = document.createElement('audio')

  constructor(private audioSrc: string) {
    this.$audio.src = this.audioSrc
  }

  apply(player: Player) {
    player.on(['play', 'pause'], ({ type }) => {
      //@ts-ignore
      this.$audio[type]()
      this.$audio.currentTime = player.currentTime
    })

    player.on(['seeking', 'error'], () => {
      this.$audio.pause()
    })

    const loading = $.cls('loading')
    player.on(() => {
      if (player.$root.classList.contains(loading)) {
        this.$audio.pause()
      }
    })

    player.on('seeked', () => {
      this.$audio.currentTime = player.currentTime
      if (player.isPlaying) this.$audio.play()
    })

    player.on('volumechange', () => {
      this.$audio.volume = player.volume
    })

    player.on('ratechange', () => {
      this.$audio.playbackRate = player.playbackRate
    })
    // .... other
  }

  destroy() {
    this.$audio.remove()
    URL.revokeObjectURL(this.$audio.src)
  }
}

```

**Comment by namdevel** [comment]
@shiyiya because the audio and video hosted in separate files

**Comment by namdevel** [comment]
@shiyiya and why i seeing this, player.context.ui.menu.notice is not a function

**Comment by shiyiya** [comment]
> @shiyiya and why i seeing this, player.context.ui.menu.notice is not a function

use latest version `@oplayer/core` `@oplayer/ui`

**Comment by namdevel** [comment]
yes, here is my code
```html
<div id="oplayer"></div>
<script src="https://cdn.jsdelivr.net/npm/@oplayer/core@latest/dist/index.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/@oplayer/ui@latest/dist/index.core.js"></script>
<script src="https://cdn.jsdelivr.net/npm/@oplayer/core@latest/dist/index.ui.js"></script>

<script>
  const audio = new Audio("https://ngewibu.test/stream/audio/1/12771946.mp4");
  audio.load();
  const UIPlayer = OPlayer.ui({
    fullscreen: true,
    coverButton: true,
    miniProgressBar: true,
    autoFocus: true,
    forceLandscapeOnFullscreen: true,
    screenshot: true,
    pictureInPicture: false,
    showControls: "always",
    keyboard: { focused: true, global: false },
    settings: ["loop"],
    theme: { primaryColor: "#ffc107" },
    speeds: ["2.0", "1.75", "1.25", "1.0", "0.75", "0.5"],
    slideToSeek: "none",
    controlBar: false, // | { back:  'always' | 'fullscreen' } // appbar
    topSetting: false, //show setting icon on appbar
    /* Default value, Optional */

    /* Custom options */
    subtitle: {
      color: "white",
      fontSize: 30,
      fontFamily: "",
      source: [
        {
          name: "Bahasa Indonesia",
          default: true,
          src: "sub.vtt",
        },
      ],
    },
    thumbnails: {
      number: 100,
      src: "https://oplayer.vercel.app/thumbnails.jpg",
    },
  });

  const vplayer = OPlayer.make("#oplayer", {
    source: {
      src: "https://ngewibu.test/stream/videos/1/12771946.mp4",
      poster:
        "https://ngewibu.test/static/img/ogv/5d04814e48ef572e97cb18bbe928bbf8086c38cf.png@720w_405h_1e_1c_90q.webp",
      title: "Ngewibu.test",
    },
  }).use([UIPlayer]);
  vplayer.create();
  vplayer.context.ui.menu.notice("test hello"); // test notice
</script>
<script>
  function playMedia() {
    audio.play();
  }

  function pauseMedia() {
    audio.pause();
  }
  function seekMedia() {
    audio.currentTime = vplayer.currentTime;
  }
  // Keep audio and video in sync
  function syncMedia() {
    audio.currentTime = vplayer.currentTime;
  }

  // Event listeners
  vplayer.on("play", playMedia);
  vplayer.on("durationchange", syncMedia);
  vplayer.on("pause", pauseMedia);
  vplayer.on("seeked", seekMedia);
</script>

```

**Comment by shiyiya** [comment]
```html

<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Document</title>
  </head>
  <body>
    <div id="oplayer"></div>
    <script src="https://cdn.jsdelivr.net/npm/@oplayer/core@latest/dist/index.min.js"></script>
    <script src="https://cdn.jsdelivr.net/npm/@oplayer/ui@latest/dist/index.min.js"></script>
    <script src="https://cdn.jsdelivr.net/npm/@oplayer/hls@latest/dist/index.hls.js"></script>

    <script>
      const audio = new Audio('https://ngewibu.test/stream/audio/1/12771946.mp4')
      audio.load()
      const UI = OUI({
        fullscreen: true,
        coverButton: true,
        miniProgressBar: true,
        autoFocus: true,
        forceLandscapeOnFullscreen: true,
        screenshot: true,
        pictureInPicture: false,
        showControls: 'always',
        keyboard: { focused: true, global: false },
        settings: ['loop'],
        theme: { primaryColor: '#ffc107' },
        speeds: ['2.0', '1.75', '1.25', '1.0', '0.75', '0.5'],
        slideToSeek: 'none',
        controlBar: false, // | { back:  'always' | 'fullscreen' } // appbar
        topSetting: false, //show setting icon on appbar
        /* Default value, Optional */

        /* Custom options */
        subtitle: {
          color: 'white',
          fontSize: 30,
          fontFamily: '',
          source: [
            {
              name: 'Bahasa Indonesia',
              default: true,
              src: 'sub.vtt',
            },
          ],
        },
        thumbnails: {
          number: 100,
          src: 'https://oplayer.vercel.app/thumbnails.jpg',
        },
      })

      const vplayer = OPlayer.make('#oplayer', {
        source: {
          src: 'https://ngewibu.test/stream/videos/1/12771946.mp4',
          poster: 'https://ngewibu.test/static/img/ogv/5d04814e48ef572e97cb18bbe928bbf8086c38cf.png@720w_405h_1e_1c_90q.webp',
          title: 'Ngewibu.test',
        },
      })
        .use([UI])
        .create()
      console.log(vplayer.context)
      vplayer.context.ui.notice('test hello') // test notice
    </script>
    <script>
      function playMedia() {
        audio.play()
      }

      function pauseMedia() {
        audio.pause()
      }
      function seekMedia() {
        audio.currentTime = vplayer.currentTime
      }
      // Keep audio and video in sync
      function syncMedia() {
        audio.currentTime = vplayer.currentTime
      }

      // Event listeners
      vplayer.on('play', playMedia)
      vplayer.on('durationchange', syncMedia)
      vplayer.on('pause', pauseMedia)
      vplayer.on('seeked', seekMedia)
    </script>
  </body>
</html>

```

**Comment by namdevel** [comment]
thank you @shiyiya 
can this player add watermark image ?
like in top left or whenever custom position ?

**Comment by shiyiya** [comment]
> thank you @shiyiya can this player add watermark image ? like in top left or whenever custom position ?

```
document.createlm(img)
img.src =''
img.style.position= ''
vplayer.$root.appendChild(img)

**Comment by namdevel** [comment]
thank you

```javascript
var img = document.createElement("img");
img.src = "layer.png";
img.style.position = "absolute";
img.style.top = "10px";
img.style.right = "10px";
img.style.width = "200px";
img.style.height = "auto";
vplayer.$root.appendChild(img);
```

**Comment by namdevel** [comment]
i hope you can add watermark in Oplayer options in future 
I wanted to take a moment to express my sincerest thanks for your incredible work.

**Comment by shiyiya** [comment]
> i hope you can add watermark in Oplayer options in future I wanted to take a moment to express my sincerest thanks for your incredible work.

I will consider it, thank u 😄

**Comment by namdevel** [comment]
![Screenshot_76](https://github.com/shiyiya/oplayer/assets/62225978/0423bd9d-be87-4785-a6fb-4ddd35941206)


i'm seeing this error while take a screenshot

**Comment by shiyiya** [comment]
> ![Screenshot_76](https://user-images.githubusercontent.com/62225978/239661111-0423bd9d-be87-4785-a6fb-4ddd35941206.png)
> 
> i'm seeing this error while take a screenshot

https://github.com/shiyiya/oplayer/blob/b86db76707fd61d66454913ffa0ccd009fa0763c/examples/standalone/main.ts#L51

**Comment by namdevel** [comment]
how to add custom slide in settings menu ?
like loop menu but custom and handle on off

**Comment by shiyiya** [comment]
> how to add custom slide in settings menu ? like loop menu but custom and handle on off

https://github.com/shiyiya/oplayer/blob/b86db76707fd61d66454913ffa0ccd009fa0763c/packages/ui/src/functions/speed.ts#L14

**Comment by namdevel** [comment]
i'm confused ^_^

**Comment by shiyiya** [comment]
player.context.ui.setting.register（）

**Comment by namdevel** [comment]
tester.html:117 Uncaught TypeError: Cannot read properties of undefined (reading 'setting')

**Comment by namdevel** [comment]
oh sorry my mistakes

**Comment by namdevel** [comment]
![Screenshot_77](https://github.com/shiyiya/oplayer/assets/62225978/5a8ce987-4a95-4ec1-824e-13366b1ca84b)

become default controls in mobile views/device

**Comment by shiyiya** [comment]
The real machine is not the same as the fake

> ![截图_77](https://user-images.githubusercontent.com/62225978/239675002-5a8ce987-4a95-4ec1-824e-13366b1ca84b.png)
> 
> 成为移动视图/设备中的默认控件

**Comment by namdevel** [comment]
i try in my real iOS device and it's same , that was appeared while in the fullscreen mode

**Comment by shiyiya** [comment]
> i try in my real iOS device and it's same , that was appeared while in the fullscreen mode

Yes, this is normal, ios can only video in full screen

**Comment by namdevel** [comment]
AAh ok, thank you

**Comment by namdevel** [comment]
![WhatsApp Image 2023-05-20 at 20 17 34](https://github.com/shiyiya/oplayer/assets/62225978/0a80041e-a20e-44fa-88ae-d4cd231b0c16)


while i try use another video player, they can show custom controls in my Device

**Comment by shiyiya** [comment]
> ![WhatsApp Image 2023-05-20 at 20 17 34](https://user-images.githubusercontent.com/62225978/239689646-0a80041e-a20e-44fa-88ae-d4cd231b0c16.jpeg)
> 
> while i try use another video player, they can show custom controls in my Device

Yes, there are some situations where custom controls can be displayed, such as iOS Chrome, Safari, and iPad. However, I do not have a corresponding device and cannot test this, so iOS uses its own full screen as a unified solution 😢

**Comment by shiyiya** [comment]
> ![WhatsApp Image 2023-05-20 at 20 17 34](https://user-images.githubusercontent.com/62225978/239689646-0a80041e-a20e-44fa-88ae-d4cd231b0c16.jpeg)
> 
> while i try use another video player, they can show custom controls in my Device

try `npm i oplayer/core@1.2.26-beta.1`

**Comment by namdevel** [comment]
is that available on jsdeliver version ?
because i'm not use npm

**Comment by shiyiya** [comment]
<script src="https://cdn.jsdelivr.net/npm/@oplayer/core@latest/dist/index.min.js"></script>

replace 
    <script src="https://cdn.jsdelivr.net/npm/@oplayer/core@1.2.26-beta.1/dist/index.min.js"></script>

**Comment by namdevel** [comment]
how to add multiple thumbnails

image 1
image 2


x : 10
y: 10

x_sizes: 160
y_sizes: 90

**Comment by shiyiya** [comment]
> how to add multiple thumbnails
> 
> image 1 image 2
> 
> x : 10 y: 10
> 
> x_sizes: 160 y_sizes: 90

https://oplayer.vercel.app/ui/#thumbnails
```
thumbnails = {
  src: ['image1','image2'],
  x: 10,
  y: 6,
  number: 60
}
```

**Comment by namdevel** [comment]
not work, the thumbnails not same with video time

**Comment by shiyiya** [comment]
> not work, the thumbnails not same with video time

Can you provide an example?

**Comment by namdevel** [comment]
https://i0.hdslb.com/bfs/videoshot/n230403er32xsq8sww4izs2q666shb65.jpg
https://i0.hdslb.com/bfs/videoshot/n230403er32xsq8sww4izs2q666shb65-1.jpg

**Comment by shiyiya** [comment]
403
If you can provide accessible videos and pictures, I can try to fix the error tomorrow

**Comment by namdevel** [comment]
forbidden access from videoplayer or from url ?

**Comment by shiyiya** [comment]
> 禁止从 videoplayer 或 url 访问？

yea

**Comment by shiyiya** [comment]
> not work, the thumbnails not same with video time

Of course, it is possible that the number of thumbnails generated is small and the gaps are too large so they are not that accurate

**Comment by namdevel** [comment]
![thumb](https://github.com/shiyiya/oplayer/assets/62225978/18d18bd0-1c70-4c66-98e9-b5663bd02f58)
![thumb-1](https://github.com/shiyiya/oplayer/assets/62225978/9acd1be7-2f3b-447a-adf9-77daf7b5a0ea)

video duration is 24:00 minutes

**Comment by shiyiya** [comment]
> video duration is 24:00 minutes

thumbnails = {
  src: ['image1','image2'],
  x: 10,
  y: 10,
  number: 192
}

**Comment by shiyiya** [comment]
Can you provide a complete example, fork template, add video address and thumbnails configuration
https://codesandbox.io/s/oplayer-m4inym

**Comment by namdevel** [comment]
![Screenshot_78](https://github.com/shiyiya/oplayer/assets/62225978/036d8ff6-f159-40f8-9fff-4138de58622f)

i'ts normal ?

**Comment by shiyiya** [comment]
I'm not sure if it's normal, it depends on the number of thumbnails you generate, for example, if the video is 10 minutes long and 2 thumbnails are generated, then it's quite inaccurate.
If your video is 24 minutes long and there are only 200 thumbnails, then the error of one thumbnail is 24*60/200=13.xxs. As long as the image exists before or after 13s of the video, it means it is normal.

**Comment by namdevel** [comment]
![Screenshot_79](https://github.com/shiyiya/oplayer/assets/62225978/e856697a-e8cd-4416-8eef-b9636af2d47d)
but on bilibili that was same, and in bilibili website player it seem like your player

**Comment by shiyiya** [comment]
OK, it is certain that it should be an issue with oplaurr.
But to fix it you need to provide videos and thumbnails 😂


> Can you provide a complete example, fork template, add video address and thumbnails configuration https://codesandbox.io/s/oplayer-m4inym

**Comment by namdevel** [comment]
i has been add

**Comment by namdevel** [comment]
hmm how to save code that i has been modified ?

**Comment by shiyiya** [comment]
> hmm how to save code that i has been modified ?

just ctrl+s and then share the link with me sometime, or just copy the link in the address bar?

**Comment by namdevel** [comment]
https://4r7q6f.csb.app/

**Comment by shiyiya** [comment]
> https://4r7q6f.csb.app/

Not this one, but a link like codesanbox.io/xxxx

**Comment by namdevel** [comment]
you can click open sandbox in bottom right

https://codesandbox.io/s/4r7q6f

**Comment by namdevel** [comment]
it's like the thumbnail is 10s faster than video

i mean the 3:53 thumnails is appeared on 3:43

**Comment by shiyiya** [comment]
i got error， Maybe I don't have access? 

MEDIA_ERR_SRC_NOT_SUPPORTED

**Comment by shiyiya** [comment]
Access Denied
You don't have permission to access "http://upos-bstar1-mirrorakam.akamaized.net/iupxcodeboss/65/hb/n230403er32xsq8sww4izs2q666shb65-1-241210110000.m4s?"

**Comment by namdevel** [comment]
it's seem your country blocked by bilibili server
you from china why bilibili blocked your country

**Comment by namdevel** [comment]
bilibili also from china

**Comment by shiyiya** [comment]
Featured 🤡

**Comment by shiyiya** [comment]
I'll fixing it tomorrow! 
I am here late at night

**Comment by namdevel** [comment]
ok, thank you

**Comment by shiyiya** [comment]
@namdevel I guess you can set the number to 200 and try

**Comment by namdevel** [comment]
![Screenshot_80](https://github.com/shiyiya/oplayer/assets/62225978/c279db79-6a07-4c78-b608-5becc53c4a73)
if i set 200, the image red arrow in thumbnails become green arrow

red arrow is 03:53 minutes

**Comment by namdevel** [comment]
this is the json about thumbnail on bilibili

```json
{
  "code": 0,
  "message": "0",
  "ttl": 1,
  "data": {
    "pv_data": "https://pic.bstarstatic.com/videoshot/n230403er32xsq8sww4izs2q666shb65.bin",
    "x_len": 10,
    "y_len": 10,
    "x_size": 160,
    "y_size": 90,
    "images": [
      "https://pic.bstarstatic.com/videoshot/n230403er32xsq8sww4izs2q666shb65.jpg",
      "https://pic.bstarstatic.com/videoshot/n230403er32xsq8sww4izs2q666shb65-1.jpg"
    ]
  }
}
```

**Comment by shiyiya** [comment]
You can try pulling the bilibili video to the end to see if the video and thumbnails match。

**Comment by namdevel** [comment]
the thumbnails match

**Comment by shiyiya** [comment]
number182， 3：53match but ed maybe not match

**Comment by shiyiya** [comment]
No idea

> not work, the thumbnails not same with video time

**Comment by shiyiya** [comment]
> how to get codecs type ?

ffmpeg

**Comment by namdevel** [comment]
ok thank you, any update about thumbnails ??

**Comment by shiyiya** [comment]
> ok thank you, any update about thumbnails ??

I don't think oplayer is wrong.

**Comment by shiyiya** [comment]
> not work, the thumbnails not same with video time

fixed!  @oplayer/ui 1.2.28
